rectools/models/candidate_ranking.py

```diff
@@ -0,0 +1,637 @@
+
+        for identifier, candgen in self.cand_gen_dict.items():
+            candidates = candgen.generate_candidates(
+                users=users,
+                dataset=dataset,
+                filter_viewed=filter_viewed,
+                for_train=for_train,
+                items_to_recommend=items_to_recommend,
+                on_unsupported_targets=on_unsupported_targets,
+            )
+
+            # Process ranks and scores as features
+            rank_col_name, score_col_name = f"{identifier}_rank", f"{identifier}_score"
+
+            candidates.rename(
+                columns={Columns.Rank: rank_col_name, Columns.Score: score_col_name},
+                inplace=True,
+                errors="ignore",
+            )
+            candidates_dfs.append(candidates)
+
+        # Merge all candidates together and process missing ranks and scores
+        all_candidates = reduce(lambda a, b: a.merge(b, how="outer", on=Columns.UserItem), candidates_dfs)
+        first_stage_results = self._process_ranks_and_scores(all_candidates)
+
+        return first_stage_results
+
+    def _process_ranks_and_scores(
+        self,
+        all_candidates: pd.DataFrame,
+    ) -> pd.DataFrame:
+        """
+        Process ranks and scores of the candidates.
+
+        Parameters
+        ----------
+        all_candidates : pd.DataFrame
+            DataFrame containing all candidates.
+
+        Returns
+        -------
+        pd.DataFrame
+            DataFrame with processed ranks and scores.
+        """
+
+        for identifier, candgen in self.cand_gen_dict.items():
+            rank_col_name, score_col_name = f"{identifier}_rank", f"{identifier}_score"
+            if candgen.keep_ranks and candgen.ranks_fillna_value is not None:
+                all_candidates[rank_col_name] = all_candidates[rank_col_name].fillna(candgen.ranks_fillna_value)
+            if candgen.keep_scores and candgen.scores_fillna_value is not None:
+                all_candidates[score_col_name] = all_candidates[score_col_name].fillna(candgen.scores_fillna_value)
+
+        return all_candidates
+
+    def recommend(
+        self,
+        users: ExternalIds,
+        dataset: Dataset,
+        k: int,
+        filter_viewed: bool,
+        items_to_recommend: tp.Optional[ExternalIds] = None,
+        add_rank_col: bool = True,
+        on_unsupported_targets: ErrorBehaviour = "raise",
+    ) -> pd.DataFrame:
+        self._check_is_fitted()
+        self._check_k(k)
+        # TODO: we should probably add an argument to force fitting again on new dataset
+        try:
+            candidates = self._get_candidates_from_first_stage(
+                users=users,
+                dataset=dataset,
+                filter_viewed=filter_viewed,
+                items_to_recommend=items_to_recommend,
+                for_train=False,
+                on_unsupported_targets=on_unsupported_targets,
+            )
+        except NotFittedError:
```

Comment:
Why do we consider 2 possible valid scenarios here (models are fitted and not fitted)?

Reply:
In real service it is common that we use a fitted reranker but want to fit first stage on fresh data for inference.
I also have a todo comment here that "we should probably add an argument to force fitting again on new dataset"
Maybe we also should pass an argument to `fit` if we want to skip fitting first-stage on full dataset after reranker was fitted. Again - this model can be used in real services where it is not needed to fit first-stage on full dataset after reranker was fitted.

Reply:
Added flags to both fit and recommend. Gonna add tests in next PR